Improve CreateVolume performance in topology aware environments

shalini-b opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**: This PR uses the recent topology feature enhancements to improve CreateVolume performance in topology aware environments. 
CreateVolume calls with accessibility requirements in a scaled environment suffer a significant overhead of traversing through all the nodeVMs morefs in the cluster to figure out which NodeVM belongs to the topology requirement every single time CreateVolume is called. The earlier implementation also created a TagManager for every CreateVolume call which can be a bottleneck in scaled environments. This PR eliminates both these problems.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Testing done**:
Vanilla E2E testing - 
```
Ran 47 of 211 Specs in 8772.275 seconds
SUCCESS! -- 47 Passed | 0 Failed | 0 Pending | 164 Skipped
PASS
```

Manual testing - 
1. CreateVolume using SC with:
- no accessibility requirements set
- improved-volume-topology FSS set to true

```
2021-06-28T20:40:54.687Z	INFO	vanilla/controller.go:659	CreateVolume: called with args {Name:pvc-10bf8459-1b32-4bd2-bb2b-66ee5432a1a7 CapacityRange:required_bytes:1073741824  VolumeCapabilities:[mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > ] Parameters:map[] Secrets:map[] VolumeContentSource:<nil> AccessibilityRequirements:requisite:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > > preferred:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > >  XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}	{"TraceId": "64e3a864-5559-4aa2-b1b5-791361ee258d"}
2021-06-28T20:40:54.732Z	INFO	csinodetopology/csinodetopology.go:491	Obtained list of nodeVMs [VirtualMachine:vm-72 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203b2c2-1dbe-3cf2-77f1-866dc7846587, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]] VirtualMachine:vm-74 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203bb1a-e50e-e2d0-cf5c-cab6995f2003, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]] VirtualMachine:vm-75 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203ef94-d938-7ef4-9ab5-8131432cc97b, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]] VirtualMachine:vm-73 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 42038285-6cd1-cce3-74d2-5f06870dd0da, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]]]	{"TraceId": "64e3a864-5559-4aa2-b1b5-791361ee258d"
2021-06-28T20:40:54.827Z	INFO	csinodetopology/csinodetopology.go:506   Obtained shared datastores: [Datastore: Datastore:datastore-41, datastore URL: ds:///vmfs/volumes/7cd882db-5098ce6f/ Datastore: Datastore:datastore-42, datastore URL: ds:///vmfs/volumes/71959bc3-a935c5ee/ Datastore: Datastore:datastore-57, datastore URL: ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/]	{"TraceId": "64e3a864-5559-4aa2-b1b5-791361ee258d"}
2021-06-28T20:40:54.837Z	DEBUG	common/vsphereutil.go:170	vSphere CSI driver creating volume pvc-10bf8459-1b32-4bd2-bb2b-66ee5432a1a7 with create spec (*types.CnsVolumeCreateSpec)(0xc000665700)({
 DynamicData: (types.DynamicData) {
 },
 Name: (string) (len=40) "pvc-10bf8459-1b32-4bd2-bb2b-66ee5432a1a7",
 VolumeType: (string) (len=5) "BLOCK",
 Datastores: ([]types.ManagedObjectReference) (len=3 cap=4) {
  (types.ManagedObjectReference) Datastore:datastore-41,
  (types.ManagedObjectReference) Datastore:datastore-42,
  (types.ManagedObjectReference) Datastore:datastore-57
 },
 Metadata: (types.CnsVolumeMetadata) {
  DynamicData: (types.DynamicData) {
  },
  ContainerCluster: (types.CnsContainerCluster) {
   DynamicData: (types.DynamicData) {
   },
   ClusterType: (string) (len=10) "KUBERNETES",
   ClusterId: (string) (len=34) "vSAN-FVT-Cluster1623287378.1728246",
   VSphereUser: (string) (len=27) "Administrator@vsphere.local",
   ClusterFlavor: (string) (len=7) "VANILLA",
   ClusterDistribution: (string) (len=11) "CSI-Vanilla"
  },
  EntityMetadata: ([]types.BaseCnsEntityMetadata) <nil>,
  ContainerClusterArray: ([]types.CnsContainerCluster) (len=1 cap=1) {
   (types.CnsContainerCluster) {
    DynamicData: (types.DynamicData) {
    },
    ClusterType: (string) (len=10) "KUBERNETES",
    ClusterId: (string) (len=34) "vSAN-FVT-Cluster1623287378.1728246",
    VSphereUser: (string) (len=27) "Administrator@vsphere.local",
    ClusterFlavor: (string) (len=7) "VANILLA",
    ClusterDistribution: (string) (len=11) "CSI-Vanilla"
   }
  }
 },
 BackingObjectDetails: (*types.CnsBlockBackingDetails)(0xc0004ebcb0)({
  CnsBackingObjectDetails: (types.CnsBackingObjectDetails) {
   DynamicData: (types.DynamicData) {
   },
   CapacityInMb: (int64) 1024
  },
  BackingDiskId: (string) "",
  BackingDiskUrlPath: (string) ""
 }),
 Profile: ([]types.BaseVirtualMachineProfileSpec) <nil>,
 CreateSpec: (types.BaseCnsBaseCreateSpec) <nil>,
 VolumeSource: (types.BaseCnsVolumeSource) <nil>
})
	{"TraceId": "64e3a864-5559-4aa2-b1b5-791361ee258d"}
2021-06-28T20:40:59.595Z	INFO	volume/manager.go:373	CreateVolume: VolumeName: "pvc-10bf8459-1b32-4bd2-bb2b-66ee5432a1a7", opId: "1f7c5e3b"	{"TraceId": "64e3a864-5559-4aa2-b1b5-791361ee258d"}
2021-06-28T20:40:59.604Z	INFO	volume/util.go:324	Volume created successfully. VolumeName: "pvc-10bf8459-1b32-4bd2-bb2b-66ee5432a1a7", volumeID: "23d1a4c3-85aa-4f8d-a6cc-c1d2523367b5"	{"TraceId": "64e3a864-5559-4aa2-b1b5-791361ee258d"}
```

2. CreateVolume with:
-  accessibility requirements set in SC
- volumeBinding as WaitForFirstConsumer
- improved-volume-topology FSS set to true

```
2021-06-28T22:54:13.245Z	INFO	vanilla/controller.go:659	CreateVolume: called with args {Name:pvc-eb86efd1-03f5-4633-9116-860b8a913de1 CapacityRange:required_bytes:1073741824  VolumeCapabilities:[mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > ] Parameters:map[] Secrets:map[] VolumeContentSource:<nil> AccessibilityRequirements:requisite:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > > preferred:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > >  XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.246Z	DEBUG	csinodetopology/csinodetopology.go:437	Get shared datastores with topologyRequirement: requisite:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > > preferred:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > > 	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.246Z	DEBUG	csinodetopology/csinodetopology.go:447	Using preferred topology	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.246Z	DEBUG	csinodetopology/csinodetopology.go:483	Getting list of nodeVMs for topology segments map[topology.kubernetes.io/region:region-1 topology.kubernetes.io/zone:zone-a]	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.280Z	INFO	csinodetopology/csinodetopology.go:491	Obtained list of nodeVMs [VirtualMachine:vm-74 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203bb1a-e50e-e2d0-cf5c-cab6995f2003, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]] VirtualMachine:vm-75 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203ef94-d938-7ef4-9ab5-8131432cc97b, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]] VirtualMachine:vm-73 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 42038285-6cd1-cce3-74d2-5f06870dd0da, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]] VirtualMachine:vm-72 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203b2c2-1dbe-3cf2-77f1-866dc7846587, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]]]	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.357Z	INFO	csinodetopology/csinodetopology.go:506	Obtained shared datastores: [Datastore: Datastore:datastore-41, datastore URL: ds:///vmfs/volumes/7cd882db-5098ce6f/ Datastore: Datastore:datastore-42, datastore URL: ds:///vmfs/volumes/71959bc3-a935c5ee/ Datastore: Datastore:datastore-57, datastore URL: ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/]	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.358Z	DEBUG	vanilla/controller.go:463	Shared datastores [[Datastore: Datastore:datastore-41, datastore URL: ds:///vmfs/volumes/7cd882db-5098ce6f/ Datastore: Datastore:datastore-42, datastore URL: ds:///vmfs/volumes/71959bc3-a935c5ee/ Datastore: Datastore:datastore-57, datastore URL: ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/]] retrieved for topologyRequirement [requisite:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > > preferred:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > > ] with datastoreTopologyMap [+map[ds:///vmfs/volumes/71959bc3-a935c5ee/:[map[topology.kubernetes.io/region:region-1 topology.kubernetes.io/zone:zone-a]] ds:///vmfs/volumes/7cd882db-5098ce6f/:[map[topology.kubernetes.io/region:region-1 topology.kubernetes.io/zone:zone-a]] ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/:[map[topology.kubernetes.io/region:region-1 topology.kubernetes.io/zone:zone-a]]]]	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.358Z	DEBUG	vanilla/controller.go:345	filterDatastores: dsMap map[ds:///vmfs/volumes/60c1645f-fa3b0638-169d-020a03fcd08b/:Datastore: Datastore:datastore-37, datastore URL: ds:///vmfs/volumes/60c1645f-fa3b0638-169d-020a03fcd08b/ ds:///vmfs/volumes/60c16461-a81cb63c-289d-020a03fcd08b/:Datastore: Datastore:datastore-38, datastore URL: ds:///vmfs/volumes/60c16461-a81cb63c-289d-020a03fcd08b/ ds:///vmfs/volumes/60c16462-a66a70e8-39dd-020a037fedaf/:Datastore: Datastore:datastore-35, datastore URL: ds:///vmfs/volumes/60c16462-a66a70e8-39dd-020a037fedaf/ ds:///vmfs/volumes/60c16463-5f552e94-1dab-020a037fedaf/:Datastore: Datastore:datastore-36, datastore URL: ds:///vmfs/volumes/60c16463-5f552e94-1dab-020a037fedaf/ ds:///vmfs/volumes/60c16466-fe09e9e0-09a0-020a0359dbb0/:Datastore: Datastore:datastore-32, datastore URL: ds:///vmfs/volumes/60c16466-fe09e9e0-09a0-020a0359dbb0/ ds:///vmfs/volumes/60c16467-a36988d4-02d0-020a0359dbb0/:Datastore: Datastore:datastore-33, datastore URL: ds:///vmfs/volumes/60c16467-a36988d4-02d0-020a0359dbb0/ ds:///vmfs/volumes/60c16468-b83601ca-b9cc-020a0378e15d/:Datastore: Datastore:datastore-39, datastore URL: ds:///vmfs/volumes/60c16468-b83601ca-b9cc-020a0378e15d/ ds:///vmfs/volumes/60c1646a-772903da-1653-020a0378e15d/:Datastore: Datastore:datastore-40, datastore URL: ds:///vmfs/volumes/60c1646a-772903da-1653-020a0378e15d/ ds:///vmfs/volumes/71959bc3-a935c5ee/:Datastore: Datastore:datastore-42, datastore URL: ds:///vmfs/volumes/71959bc3-a935c5ee/ ds:///vmfs/volumes/7cd882db-5098ce6f/:Datastore: Datastore:datastore-41, datastore URL: ds:///vmfs/volumes/7cd882db-5098ce6f/ ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/:Datastore: Datastore:datastore-57, datastore URL: ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/] sharedDatastores [Datastore: Datastore:datastore-41, datastore URL: ds:///vmfs/volumes/7cd882db-5098ce6f/ Datastore: Datastore:datastore-42, datastore URL: ds:///vmfs/volumes/71959bc3-a935c5ee/ Datastore: Datastore:datastore-57, datastore URL: ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/]	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.358Z	DEBUG	vanilla/controller.go:354	filterDatastores: filteredDatastores [Datastore: Datastore:datastore-41, datastore URL: ds:///vmfs/volumes/7cd882db-5098ce6f/ Datastore: Datastore:datastore-42, datastore URL: ds:///vmfs/volumes/71959bc3-a935c5ee/ Datastore: Datastore:datastore-57, datastore URL: ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/]	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.365Z	DEBUG	common/vsphereutil.go:170	vSphere CSI driver creating volume pvc-eb86efd1-03f5-4633-9116-860b8a913de1 with create spec (*types.CnsVolumeCreateSpec)(0xc000b75600)({
 DynamicData: (types.DynamicData) {
 },
 Name: (string) (len=40) "pvc-eb86efd1-03f5-4633-9116-860b8a913de1",
 VolumeType: (string) (len=5) "BLOCK",
 Datastores: ([]types.ManagedObjectReference) (len=3 cap=4) {
  (types.ManagedObjectReference) Datastore:datastore-41,
  (types.ManagedObjectReference) Datastore:datastore-42,
  (types.ManagedObjectReference) Datastore:datastore-57
 },
 Metadata: (types.CnsVolumeMetadata) {
  DynamicData: (types.DynamicData) {
  },
  ContainerCluster: (types.CnsContainerCluster) {
   DynamicData: (types.DynamicData) {
   },
   ClusterType: (string) (len=10) "KUBERNETES",
   ClusterId: (string) (len=34) "vSAN-FVT-Cluster1623287378.1728246",
   VSphereUser: (string) (len=27) "Administrator@vsphere.local",
   ClusterFlavor: (string) (len=7) "VANILLA",
   ClusterDistribution: (string) (len=11) "CSI-Vanilla"
  },
  EntityMetadata: ([]types.BaseCnsEntityMetadata) <nil>,
  ContainerClusterArray: ([]types.CnsContainerCluster) (len=1 cap=1) {
   (types.CnsContainerCluster) {
    DynamicData: (types.DynamicData) {
    },
    ClusterType: (string) (len=10) "KUBERNETES",
    ClusterId: (string) (len=34) "vSAN-FVT-Cluster1623287378.1728246",
    VSphereUser: (string) (len=27) "Administrator@vsphere.local",
    ClusterFlavor: (string) (len=7) "VANILLA",
    ClusterDistribution: (string) (len=11) "CSI-Vanilla"
   }
  }
 },
 BackingObjectDetails: (*types.CnsBlockBackingDetails)(0xc000b24780)({
  CnsBackingObjectDetails: (types.CnsBackingObjectDetails) {
   DynamicData: (types.DynamicData) {
   },
   CapacityInMb: (int64) 1024
  },
  BackingDiskId: (string) "",
  BackingDiskUrlPath: (string) ""
 }),
 Profile: ([]types.BaseVirtualMachineProfileSpec) <nil>,
 CreateSpec: (types.BaseCnsBaseCreateSpec) <nil>,
 VolumeSource: (types.BaseCnsVolumeSource) <nil>
})
	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:13.376Z	DEBUG	volume/util.go:195	Update VSphereUser from Administrator@vsphere.local to VSPHERE.LOCAL\Administrator	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:17.858Z	INFO	volume/manager.go:373	CreateVolume: VolumeName: "pvc-eb86efd1-03f5-4633-9116-860b8a913de1", opId: "1f7c85ac"	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:17.864Z	INFO	volume/util.go:324	Volume created successfully. VolumeName: "pvc-eb86efd1-03f5-4633-9116-860b8a913de1", volumeID: "8fed9309-d1ed-4902-8f14-e15786d84ee7"	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:17.864Z	DEBUG	volume/util.go:326	CreateVolume volumeId {{} "8fed9309-d1ed-4902-8f14-e15786d84ee7"} is placed on datastore "ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/"	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:17.865Z	DEBUG	vanilla/controller.go:560	Volume: 8fed9309-d1ed-4902-8f14-e15786d84ee7 is provisioned on the datastore: ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/ 	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:17.865Z	DEBUG	vanilla/controller.go:565	volumeAccessibleTopology: [map[topology.kubernetes.io/region:region-1 topology.kubernetes.io/zone:zone-a]] is selected for datastore: ds:///vmfs/volumes/vsan:524a3351f7186f14-bd9868a34c6b8ebb/ 	{"TraceId": "25ac253e-4149-416e-88b8-14e1e22abd0e"}
2021-06-28T22:54:18.558Z	INFO	vanilla/controller.go:772	ControllerPublishVolume: called with args {VolumeId:8fed9309-d1ed-4902-8f14-e15786d84ee7 NodeId:k8s-node-0158 VolumeCapability:mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER >  Readonly:false Secrets:map[] VolumeContext:map[storage.kubernetes.io/csiProvisionerIdentity:1624920530033-8081-csi.vsphere.vmware.com type:vSphere CNS Block Volume] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}	{"TraceId": "ecbb1626-7a85-44a2-89b4-834e3509cb88"}
2021-06-28T22:54:18.570Z	DEBUG	vanilla/controller.go:847	Found VirtualMachine for node:"k8s-node-0158".	{"TraceId": "ecbb1626-7a85-44a2-89b4-834e3509cb88"}
2021-06-28T22:54:18.571Z	DEBUG	common/vsphereutil.go:440	vSphere CSI driver is attaching volume: "8fed9309-d1ed-4902-8f14-e15786d84ee7" to vm: "VirtualMachine:vm-75 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203ef94-d938-7ef4-9ab5-8131432cc97b, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]]"	{"TraceId": "ecbb1626-7a85-44a2-89b4-834e3509cb88"}
2021-06-28T22:54:22.083Z	INFO	volume/manager.go:477	AttachVolume: volumeID: "8fed9309-d1ed-4902-8f14-e15786d84ee7", vm: "VirtualMachine:vm-75 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203ef94-d938-7ef4-9ab5-8131432cc97b, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]]", opId: "1f7c85c6"	{"TraceId": "ecbb1626-7a85-44a2-89b4-834e3509cb88"}
2021-06-28T22:54:22.083Z	INFO	volume/manager.go:512	AttachVolume: Volume attached successfully. volumeID: "8fed9309-d1ed-4902-8f14-e15786d84ee7", opId: "1f7c85c6", vm: "VirtualMachine:vm-75 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203ef94-d938-7ef4-9ab5-8131432cc97b, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]]", diskUUID: "6000C29b-f1e8-46b0-a92f-708982c59e03"	{"TraceId": "ecbb1626-7a85-44a2-89b4-834e3509cb88"}
2021-06-28T22:54:22.083Z	DEBUG	common/vsphereutil.go:446	Successfully attached disk 8fed9309-d1ed-4902-8f14-e15786d84ee7 to VM VirtualMachine:vm-75 [VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com, UUID: 4203ef94-d938-7ef4-9ab5-8131432cc97b, Datacenter: Datacenter [Datacenter: Datacenter:datacenter-3, VirtualCenterHost: sc2-10-186-80-158.eng.vmware.com]]. Disk UUID is 6000C29b-f1e8-46b0-a92f-708982c59e03	{"TraceId": "ecbb1626-7a85-44a2-89b4-834e3509cb88"}
2021-06-28T22:54:22.084Z	INFO	vanilla/controller.go:856	ControllerPublishVolume successful with publish context: map[diskUUID:6000c29bf1e846b0a92f708982c59e03 type:vSphere CNS Block Volume]	{"TraceId": "ecbb1626-7a85-44a2-89b4-834e3509cb88"}
```

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Improve CreateVolume performance in topology aware environments
```

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 108

svcbot-qecnsdp commented:
``` 
Build ID: 108
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 45 of 195 Specs in 12759.080 seconds
FAIL! -- 44 Passed | 1 Failed | 0 Pending | 150 Skipped
--- FAIL: TestE2E (12759.17s)
FAIL

Ginkgo ran 1 suite in 3h33m20.372533043s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-block-vanilla-pre-check-in/108/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 110

svcbot-qecnsdp commented:
``` 
Build ID: 110
Block vanilla build status: SUCCESS 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 

Ran 1 of 195 Specs in 115.129 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 194 Skipped
PASS

Ginkgo ran 1 suite in 2m32.258461882s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/csi-block-vanilla-pre-check-in/110/vsphere-csi-driver`

```

shalini-b commented on pkg/common/cns-lib/vsphere/virtualmachine.go:
```diff
@@ -337,3 +337,38 @@ func (vm *VirtualMachine) IsInZoneRegion(ctx context.Context, zoneCategoryName s
 	return strings.TrimPrefix(providerID, providerPrefix)
 }
+
+// GetSharedDatastoresForVMs returns shared datastores accessible to specified
```
No changes were made to this function. It has just been copied from `nodes.go` to `virtualmachine.go`

SandeepPissay commented:
@divyenpatel @shalini-b we discussed below things related to this PR:
1. CSI code need to be generic and should not have k8s specific code, and any CO specific code should be behind a COAgnostic interface so that its easy to support non k8s clusters in the future. This is also our design guideline. We discussed this in the context of this PR and we need some changes in the PR to be inline with the guideline.
2. @shalini-b mentioned that we have added a similar code for idempotency improvement that @RaunakShah is leading. Shalini, can you point us to the code where the above guideline is not followed?
3. We discussed a new design or an optimized way to find the shared datastores during volume provisioning operation in a topology aware deployment. @divyenpatel @shalini-b , please get back on this investigation and your recommendations.
4. We discussed that @shalini-b will come up with a doc that explains how volume provisioning works in CSI for different combinations of volumeBindingModes(immediate, wffc) and accessibilityConstraints(empty, zone-a, zone-b, zone-a or zone-b). We are specifically looking for the contents of preferred and requisites in the create volume request in all the 8 combinations.

shalini-b commented:
> @divyenpatel @shalini-b we discussed below things related to this PR:
> 
> 1. CSI code need to be generic and should not have k8s specific code, and any CO specific code should be behind a COAgnostic interface so that its easy to support non k8s clusters in the future. This is also our design guideline. We discussed this in the context of this PR and we need some changes in the PR to be inline with the guideline.
> 2. @shalini-b mentioned that we have added a similar code for idempotency improvement that @RaunakShah is leading. Shalini, can you point us to the code where the above guideline is not followed?
> 3. We discussed a new design or an optimized way to find the shared datastores during volume provisioning operation in a topology aware deployment. @divyenpatel @shalini-b , please get back on this investigation and your recommendations.
> 4. We discussed that @shalini-b will come up with a doc that explains how volume provisioning works in CSI for different combinations of volumeBindingModes(immediate, wffc) and accessibilityConstraints(empty, zone-a, zone-b, zone-a or zone-b). We are specifically looking for the contents of preferred and requisites in the create volume request in all the 8 combinations.

Addressed most of the review comments:
1. Moved all k8s specific code from `node.go` and `controller.go` into the COCommon orchestrator. Retained the topology service interface as it was already CO agnostic. 
2. Pointed out the k8s specific code to Raunak and he agreed to change it.
3. @divyenpatel & I had a discussion on how we can improve the computation of shared datastores for CreateVolume and this is the conclusion we came to:
- Directly use Tagmanager with tags to find shared datastores
```
Cons:
- If the same tag is applied on entities not participating in the k8s clusters, it will create a problem while computing shared datastores.
- No SOAP client available for TagManager, so we need to use a REST client which has a rate-limit issue. The rate limiting will be a problem when we have multiple k8s clusters.
```

- For every node, cache the datastores accessible to it at regular intervals
```
Cons:
- In non-topology setups, if VM migrates then our cache will be stale
- If a customer adds/removes a datastore, CSI might use stale cache to provision volumes which will create problems during attach
```

- Current approach: For every node, fetch accessible datatstores and compute the shared datastores among the given the list of nodeVMs
```
Cons:
- We fetch the datastores from the VC inventory, so data is not stale even if VM migrates to another host or if datastores get added/removed.
- This is the approach followed in non-topology aware setups as well to find shared datastores among all nodes in the k8s cluster
```
Also, the function `GetSharedDatastoresForVMs` is common to CreateVolume called in topology and non-topology setups. Therefore, let us keep that discussion separate from this PR. I will raise an issue to track this conversation.

4. I will put up the details of this in a separate comment.
5. Removed double caching while computing the nodeVMs which match the topology requirement. Used the informer cache instead. This method is equally efficient as the map approach I had earlier.

shalini-b commented:
> @divyenpatel @shalini-b we discussed below things related to this PR:
> 
> 1. CSI code need to be generic and should not have k8s specific code, and any CO specific code should be behind a COAgnostic interface so that its easy to support non k8s clusters in the future. This is also our design guideline. We discussed this in the context of this PR and we need some changes in the PR to be inline with the guideline.
> 2. @shalini-b mentioned that we have added a similar code for idempotency improvement that @RaunakShah is leading. Shalini, can you point us to the code where the above guideline is not followed?
> 3. We discussed a new design or an optimized way to find the shared datastores during volume provisioning operation in a topology aware deployment. @divyenpatel @shalini-b , please get back on this investigation and your recommendations.
> 4. We discussed that @shalini-b will come up with a doc that explains how volume provisioning works in CSI for different combinations of volumeBindingModes(immediate, wffc) and accessibilityConstraints(empty, zone-a, zone-b, zone-a or zone-b). We are specifically looking for the contents of preferred and requisites in the create volume request in all the 8 combinations.

Reply to 4. CreateVolume calls with strict topology enabled in all 6 scenarios with PV describe output - https://gist.github.com/shalini-b/a9135720256a6dccae263210703eeb63
CreateVolume calls with strict topology disabled in all 6 scenarios with PV describe output - https://gist.github.com/shalini-b/46dec5f1eb215a57885416246be76134

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: Parallel: images 

```

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 47 of 211 Specs in 8772.275 seconds
SUCCESS! -- 47 Passed | 0 Failed | 0 Pending | 164 Skipped
PASS

Ginkgo ran 1 suite in 2h26m23.137974643s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla/Results/612/vsphere-csi-driver`

```

divyenpatel commented:
/ok-to-test

shalini-b commented on pkg/csi/service/common/commonco/k8sorchestrator/topology.go:
```diff
@@ -0,0 +1,474 @@
+				log.Errorf("failed to create an informer for CSINodeTopology instances. Error: %+v", err)
+				return nil, err
+			}
+
```
Log message is already present https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/1029/files#diff-38cf2065c11a29c2498d29b2007cfa77bab0ce8124a33d1b6a0200175b5985bbR153

shalini-b commented on pkg/common/unittestcommon/utils.go:
```diff
@@ -97,6 +98,16 @@ func (c *FakeK8SOrchestrator) ClearFakeAttached(ctx context.Context, volumeID st
 
+func (c *FakeK8SOrchestrator) InitTopologyServiceInController(ctx context.Context) (
+	commoncotypes.ControllerTopologyService, error) {
+	return &mockControllerVolumeTopology{}, nil
```
Done

shalini-b commented on pkg/common/unittestcommon/utils.go:
```diff
@@ -97,6 +98,16 @@ func (c *FakeK8SOrchestrator) ClearFakeAttached(ctx context.Context, volumeID st
+
+func (c *FakeK8SOrchestrator) InitTopologyServiceInNode(ctx context.Context) (commoncotypes.NodeTopologyService,
+	error) {
+	return &mockNodeVolumeTopology{}, nil
```
Done

shalini-b commented on pkg/csi/service/common/commonco/k8sorchestrator/topology.go:
```diff
@@ -0,0 +1,474 @@
+			log.Warnf("Received unidentified object - %+v", event.Object)
+			continue
+		}
+		if csiNodeTopologyInstance.Name != info.NodeName {
```
This check is still required. I tested this change. The field selector works in different ways for different entities.

shalini-b commented on pkg/csi/service/common/commonco/k8sorchestrator/topology.go:
```diff
@@ -0,0 +1,474 @@
+				nodeMgr:                 nodeManager,
+				csiNodeTopologyInformer: *crInformer,
+			}
+			log.Infof("Topology service initiated successfully")
```
Changed it

shalini-b commented on pkg/common/unittestcommon/types.go:
```diff
@@ -60,3 +64,16 @@ type MockVolumeMigrationService interface {
+type mockControllerVolumeTopology struct{}
+type mockNodeVolumeTopology struct{}
+
+func (nodeTopology *mockNodeVolumeTopology) GetNodeTopologyLabels(ctx context.Context, info *commoncotypes.NodeInfo) (
```
Done

shalini-b commented on pkg/common/unittestcommon/types.go:
```diff
@@ -60,3 +64,16 @@ type MockVolumeMigrationService interface {
 }
+
+type mockControllerVolumeTopology struct{}
+type mockNodeVolumeTopology struct{}
```
Done

shalini-b commented on pkg/common/unittestcommon/types.go:
```diff
@@ -60,3 +64,16 @@ type MockVolumeMigrationService interface {
 	volumeOperationRequestMap map[string]*cnsvolumeoperationrequest.VolumeOperationRequestDetails
 }
+
+type mockControllerVolumeTopology struct{}
```
Done

divyenpatel commented on pkg/csi/service/common/commonco/k8sorchestrator/topology.go:
```diff
@@ -0,0 +1,475 @@
+}
+
+// GetNodeTopologyLabels uses the CSINodeTopology CR to retrieve topology information of a node.
+func (volTopology *nodeVolumeTopology) GetNodeTopologyLabels(ctx context.Context, info *commoncotypes.NodeInfo) (
```
info => nodeInfo

shalini-b replied:
Done

divyenpatel reviewed (approved):
/approve

shalini-b commented:
Follow-up PRs to handle strict topology changes, resolving node affinity issues and to introduce topology categories will be raised after this PR is merged. Trying to avoid creating large changes in one PR.

chethanv28 commented on pkg/common/cns-lib/vsphere/virtualmachine.go:
```diff
@@ -337,3 +337,38 @@ func (vm *VirtualMachine) IsInZoneRegion(ctx context.Context, zoneCategoryName s
+
+// GetSharedDatastoresForVMs returns shared datastores accessible to specified
+// nodeVMs list.
+func GetSharedDatastoresForVMs(ctx context.Context, nodeVMs []*VirtualMachine) ([]*DatastoreInfo, error) {
```
Have we unit tested this function with empty nodeVMs[] ? 
If not, are the callers of this function checking for length of `sharedDatastores` returned by this function to avoid NPE

shalini-b replied:
yes the callers check for length of `sharedDatastores` and appropriately error out.

chethanv28 commented on pkg/csi/service/common/commonco/k8sorchestrator/topology.go:
```diff
@@ -0,0 +1,475 @@
+			// Get in cluster config for client to API server.
+			config, err := k8s.GetKubeConfig(ctx)
+			if err != nil {
+				log.Errorf("failed to get kubeconfig with error: %v", err)
```
If yes, please make sure all the logs are using the same logger method

shalini-b replied:
`LogNewErrorCodef` is used to return a certain error code along with the message. The other function `LogNewErrorf` creates a new error and logs it if we give a message as argument. It is not absolutely necessary to use these only. I already have an error here, I am just logging it and sending the original error back to the caller.

chethanv28 commented on pkg/csi/service/common/commonco/k8sorchestrator/topology.go:
```diff
@@ -0,0 +1,475 @@
+
+// getCSINodeTopologyWatchTimeoutInMin returns the timeout for watching
+// on CSINodeTopology instances for any updates.
+// If environment variable NODEGETINFO_WATCH_TIMEOUT_MINUTES is set and
```
`NODEGETINFO_WATCH_TIMEOUT_MINUTES` is not seen in any of deployment yamls. Can you confirm why ?

shalini-b replied:
I have a default value set for it in the code like a lot of our other env variables. If customer wants to increase the value as his/her node takes a longer time to come up, then they can include this env var in their deployment YAML with an increased value.

chethanv28 replied:
Then, please include the `NODEGETINFO_WATCH_TIMEOUT_MINUTES` in deployment yaml with default values

shalini-b replied:
Done

shalini-b replied:
I used the latest YAML to deploy the daemonset after these changes and it worked.
Logs:
```
2021-07-28T19:23:12.848Z	INFO	logger/logger.go:41	Setting default log level to :"DEVELOPMENT"
2021-07-28T19:23:12.851Z	INFO	vsphere-csi/main.go:57	Version : v2.1.0-rc.1-959-gdf2e6d7c-dirty	{"TraceId": "930b8e91-d294-4d95-9000-b81f4f808ac5"}
2021-07-28T19:23:12.851Z	DEBUG	commonco/utils.go:102	Container orchestrator init params: {InternalFeatureStatesConfigInfo:{Name:internal-feature-states.csi.vsphere.vmware.com Namespace:vmware-system-csi} ServiceMode:node}	{"TraceId": "930b8e91-d294-4d95-9000-b81f4f808ac5"}
2021-07-28T19:23:12.888Z	INFO	logger/logger.go:41	Setting default log level to :"DEVELOPMENT"
2021-07-28T19:23:12.888Z	INFO	k8sorchestrator/k8sorchestrator.go:152	Initializing k8sOrchestratorInstance	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:12.903Z	INFO	kubernetes/kubernetes.go:85	k8s client using in-cluster config	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:12.907Z	INFO	kubernetes/kubernetes.go:341	Setting client QPS to 100.000000 and Burst to 100.	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:13.696Z	INFO	k8sorchestrator/k8sorchestrator.go:258	New internal feature states values stored successfully: map[async-query-volume:false block-volume-snapshot:false csi-auth-check:true csi-migration:true improved-csi-idempotency:false improved-volume-topology:true online-volume-extend:true trigger-csi-fullsync:false]	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:13.728Z	INFO	k8sorchestrator/k8sorchestrator.go:178	k8sOrchestratorInstance initialized	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:13.728Z	INFO	service/driver.go:107	Configured: "csi.vsphere.vmware.com" with clusterFlavor: "VANILLA" and mode: "node"	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
time="2021-07-28T19:23:13Z" level=info msg="identity service registered"
time="2021-07-28T19:23:13Z" level=info msg="node service registered"
time="2021-07-28T19:23:13Z" level=info msg=serving endpoint="unix:///csi/csi.sock"
2021-07-28T19:23:13.740Z	INFO	k8sorchestrator/k8sorchestrator.go:484	configMapAdded: Internal feature state values from "internal-feature-states.csi.vsphere.vmware.com" stored successfully: map[async-query-volume:false block-volume-snapshot:false csi-auth-check:true csi-migration:true improved-csi-idempotency:false improved-volume-topology:true online-volume-extend:true trigger-csi-fullsync:false]	{"TraceId": "3ae6ec48-6e8d-4e71-a211-ba3a3fd1c363"}
2021-07-28T19:23:14.966Z	INFO	service/node.go:660	NodeGetInfo: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.966Z	INFO	service/node.go:682	NodeGetInfo: MAX_VOLUMES_PER_NODE is set to 0	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.966Z	INFO	kubernetes/kubernetes.go:85	k8s client using in-cluster config	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.967Z	INFO	kubernetes/kubernetes.go:341	Setting client QPS to 100.000000 and Burst to 100.	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.967Z	INFO	kubernetes/kubernetes.go:85	k8s client using in-cluster config	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.967Z	INFO	kubernetes/kubernetes.go:341	Setting client QPS to 100.000000 and Burst to 100.	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:15.027Z	INFO	k8sorchestrator/topology.go:204	Topology service initiated successfully	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:15.062Z	INFO	k8sorchestrator/topology.go:312	Timeout is set to 1 minute(s)	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:15.068Z	INFO	k8sorchestrator/topology.go:271	Got another object - "k8s-master-61"	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:15.069Z	INFO	service/node.go:753	NodeGetInfo response: node_id:"k8s-node-0378" accessible_topology:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > > 	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
```

chethanv28 commented on pkg/csi/service/vanilla/controller.go:
```diff
@@ -411,36 +421,46 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
+			}
+			tagManager, err := cnsvsphere.GetTagManager(ctx, vcenter)
+			if err != nil {
+				return nil, logger.LogNewErrorCodef(log, codes.NotFound,
```
Use `codes.Internal`

shalini-b replied:
This is the old code, I just moved it into the `else` block when FSS is not enabled. Anyways, I will change it.

chethanv28 commented on pkg/csi/service/vanilla/controller.go:
```diff
@@ -411,36 +421,46 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
+			vcenter, err := c.manager.VcenterManager.GetVirtualCenter(ctx, c.manager.VcenterConfig.Host)
 			if err != nil {
-				log.Errorf("failed to logout tagManager. err: %v", err)
+				return nil, logger.LogNewErrorCodef(log, codes.NotFound,
```
Use `codes.Internal`

shalini-b replied:
This is the old code, I just moved it into the `else` block when FSS is not enabled. Anyways, I will change it.

chethanv28 replied:
Thanks! I have create an issue to track the work involved to change at all places: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/1201

chethanv28 reviewed (approved):
/approve

chethanv28 commented:
@shalini-b  Can you run the e2e pipelines again, since all review comments are addressed now.

shalini-b commented:
> @shalini-b Can you run the e2e pipelines again, since all review comments are addressed now.

I have not made any changes to the logic after the last E2E tests were run. All of Divyen's comments were on the unittests and your comments were cosmetic too. Running E2E again is not required IMO.

chethanv28 commented:
> > @shalini-b Can you run the e2e pipelines again, since all review comments are addressed now.
> 
> I have not made any changes to the logic after the last E2E tests were run. All of Divyen's comments were on the unittests and your comments were cosmetic too. Running E2E again is not required IMO.

Since this confirms that there are no e2e test run needed, I am assuming e2e tests don't check for error codes returned by CSI(which were changed while addressing review comments)


Overall changes look good to me. Merging the PR

chethanv28 reviewed (approved):
/lgtm
